lib/ansible/module_utils/powershell/Ansible.ModuleUtils.Backup.psm1

```diff
@@ -0,0 +1,24 @@
+# Copyright (c): 2018, Dag Wieers (@dagwieers) <dag@wieers.com>
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+
+Set-StrictMode -Version 2.0
+$ErrorActionPreference = "Stop"
+
+Function Backup-File($path, $obj=@{}, $check_mode=$false) {
```

Comment:
This is more a personal choice as we haven't decided to make thing in this way but you could potentially define your function like

```
Function Backup-File($path) {
    [CmdletBinding(SupportsShouldProcess=$true)]

    ...
    Copy-Item -Path $path -Destination $backup_path
}
```

The line `[CmdletBinding(SupportsShouldProcess=$true)]` means the cmdlet now supports the standard PS cmdlet parameters like `-WhatIf`, `-Verbose` and so on. What this means the module can call this with `-WhatIf:$check_mode` and this module, any whatever it invokes, will be in in check mode for you.

Reply:
Hmmm, that's a good point. I basically assumed this to be an internal function, so a more Ansible/Python-esque interface, but since we are leaning towards more Powershell-alike APIs these days I prefer what you suggest.

On the other hand, if `-WhatIf` support means the `Backup-File` operation becomes a no-op, rather than handing `-WhatIf` over to Copy-Item I don't think we gain anything.

Even if `-WhatIf` in Copy-Item is basically a no-op I still have a hope that some day Microsoft will add proper `-WhatIf` behaviour by testing any conditions before doing a no-op.

Reply:
Ok, after reading how this works, the `-WhatIf` is passed over to any cmdlets that support it automatically. That's cool (but also terrifying maybe), but here it would actually work.

So consider it done.

Reply:
Exactly :-)

Reply:
I want to double check to make sure but this has been the behaviour I've seen before